src/sysusers/sysusers.c

```diff
@@ -567,6 +571,17 @@ static int write_temporary_shadow(const char *shadow_path, FILE **tmpfile, char
                         return errno ? -errno : -EIO;
         }
 
+        /* Append the remaining NIS entries if any */
+        while (sp) {
+                errno = 0;
+                if (putspent(sp, shadow) < 0)
+                        return errno ? -errno : -EIO;
+
+                sp = fgetspent(original);
+        }
+        if (!IN_SET(errno, 0, ENOENT))
+                return -errno;
```

Comment:
hmm, i am noticing this only now: this code isn't right actually, as errno might have been reset in the middle part, but this code assumes it's still set from the previous fgetspent()... 

This is actually borked in the passwd/group cases too.

Moreover errno needs to be reset to 0 really before we invoke fgetspent(), so that we aren't confused by any side-effect initialization by putspent().

Reply:
hmm I'm not sure to follow... do you mean that errno could be set even though all previous calls succeed ?

Reply:
yes. As a rule in C, errno is only set on failure, but it is *not* reset on success. That means that if anything inside the implementation of fgetspent() or anything it calls causes a failing syscall, but when this is not considered an error, from the code around it, then errno will remain set for good...

And don't forget that many syscalls will signal an error through errno even if nothing is really wrong. Think: access(F_OK) on some file to see if it's there. If it isn't, then errno will be set to ENOENT, even though that is likely no error at alll

Reply:
I see thanks.